File: src/permissionsPolicy.ts
/**
 * permissionsPolicy.ts — Permissions-Policy response-header
 * audit. T76.
 *
 * The Permissions-Policy header (W3C; formerly Feature-Policy)
 * gates which browser APIs the page AND any embedded iframes
 * can use. Without the header — or with overly permissive
 * directives — third-party scripts and iframes inherit ambient
 * permission to access:
 *
 *   * camera         — surveillance vector. JS can request video.
 *   * microphone     — same, audio.
 *   * geolocation    — physical-location leak.
 *   * payment        — Payment Request API (financial data).
 *   * usb / serial /
 *     midi / hid     — physical-device access; unique fingerprint
 *                      surface even if user denies.
 *   * accelerometer /
 *     gyroscope /
 *     magnetometer   — high-resolution device-motion fingerprinting,
 *                      side-channel for keystroke recovery on mobile.
 *   * display-capture / screen-wake-lock — surveillance, power-cost
 *                      attacks, screen-content recording.
 *
 * Browsers default each feature to `*` (allowed for the page
 * AND every embedded iframe) when no policy is set. Modern best
 * practice: deny by default, allow per-feature only as needed.
 *
 * Syntax: comma-separated `feature=allowlist` directives, where
 * `allowlist` is a parenthesized list:
 *   * `()`                          — deny entirely (best)
 *   * `(self)`                      — page itself only (good)
 *   * `(self "https://example.com")` — page + specific origin
 *   * `*`                           — page + all iframes (bad)
 *
 * Findings:
 *
 *   - permissions-policy.missing                warn
 *     No Permissions-Policy header. All features default to `*`
 *     — every embedded iframe can use camera/mic/geo/payment/etc.
 *
 *   - permissions-policy.allow-all-<feature>    strict
 *     A high-risk feature (camera / microphone / geolocation /
 *     payment / usb / serial / midi / hid) is set to `*` —
 *     explicitly allowing every embedded iframe. The strictest
 *     finding — operator either misunderstood the directive or
 *     forgot to restrict it.
 *
 *   - permissions-policy.invalid                warn
 *     Header value couldn't be parsed (no `=` separators, only
 *     unknown features, malformed allowlist). Browsers ignore
 *     unparseable directives → effectively missing.
 *
 * Out of scope (consistent with the response-header detector
 * family doctrine):
 *   * http pages — Permissions-Policy is honored on http but
 *     hostile networks can strip it; HSTS is the prerequisite.
 *     The detector still fires on http because there is no
 *     downside to setting the header over http (unlike Secure
 *     cookies which can't apply); but the localhost exemption
 *     stays.
 *   * Localhost / 127.0.0.1 / *.localhost — fixture-mode
 *     short-circuit, same family as hsts/xframe/referrer/cookie.
 *
 * Reads from the same `topLevelResponseHeaders` Map as hsts +
 * xFrameOptions + referrerPolicy + cookieSecurity — FIFTH
 * consumer of the shared capture path. Pattern-extraction
 * (`multiValueHeaderDetector`) remains deferred — Permissions-
 * Policy IS multi-directive (comma-separated `feature=allowlist`)
 * but each directive's allowlist parsing is feature-dependent
 * enough that extracting now would be premature. After a SIXTH
 * multi-value header detector lands (probable: full CSP), the
 * shape will be clearer.
 */

export interface PermissionsPolicyFinding {
  severity: 'strict' | 'warn';
  kind: string;
  detail: string;
  evidence: Record<string, unknown>;
}

export interface ParsedDirective {
  feature: string;
  /**
   * Raw allowlist text inside the parens (or `*` for unparenned
   * star). Empty string for `()`.
   */
  allowlistRaw: string;
  /** True iff the allowlist is `*` — every iframe allowed. */
  isAllowAll: boolean;
  /** True iff the allowlist is `()` — denied for everyone. */
  isDeny: boolean;
  /** True iff `self` appears in the allowlist. */
  hasSelf: boolean;
  /** Origins beyond `self` listed in the allowlist (quoted). */
  origins: string[];
}

export interface PermissionsPolicySnapshot {
  pageUrl: string;
  pageIsLocalhost: boolean;
  /** Raw header value, or null if absent. */
  raw: string | null;
  /** Parsed directives in declaration order. Empty if no header. */
  directives: ParsedDirective[];
  /**
   * True iff the header was present but unparseable into ANY
   * directives. Distinguishes "missing" from "garbage".
   */
  unparseable: boolean;
}

/**
 * Features the detector classifies as high-risk for the strict
 * `allow-all-<feature>` finding. Drawn from the W3C
 * Permissions-Policy spec's privacy-impacting features list +
 * payment + device-access APIs known to enable cross-origin
 * fingerprinting.
 *
 * `accelerometer`, `gyroscope`, `magnetometer` are intentionally
 * INCLUDED — even though websites use them for legitimate
 * orientation features, the side-channel literature (TouchLogger,
 * AccessLogger, etc.) proves they enable keystroke recovery on
 * mobile when allowed cross-origin.
 */
const HIGH_RISK_FEATURES: ReadonlySet<string> = new Set([
  'camera',
  'microphone',
  'geolocation',
  'payment',
  'usb',
  'serial',
  'midi',
  'hid',
  'bluetooth',
  'accelerometer',
  'gyroscope',
  'magnetometer',
  'display-capture',
  'screen-wake-lock',
]);

function isLocalhost(url: string): boolean {
  try {
    const u = new URL(url);
    const h = u.hostname;
    return h === 'localhost' || h === '127.0.0.1' || h === '::1' || h.endsWith('.localhost');
  } catch {
    return false;
  }
}

/**
 * Parse one `feature=allowlist` directive.
 *
 * Examples:
 *   `camera=()`                          → deny
 *   `camera=*`                           → allow-all
 *   `camera=(self)`                      → self only
 *   `camera=(self "https://e.com")`      → self + origin
 *   `camera=("https://e.com")`           → origin only
 *
 * Returns null on malformed input (no `=`, no value at all).
 */
function parseDirective(raw: string): ParsedDirective | null {
  const trimmed = raw.trim();
  if (!trimmed) return null;
  const eq = trimmed.indexOf('=');
  if (eq <= 0) return null;
  const feature = trimmed.slice(0, eq).trim().toLowerCase();
  if (!feature) return null;
  const value = trimmed.slice(eq + 1).trim();
  if (!value) return null;

  if (value === '*') {
    return { feature, allowlistRaw: '*', isAllowAll: true, isDeny: false, hasSelf: false, origins: [] };
  }
  // Must be parenthesized.
  if (!value.startsWith('(') || !value.endsWith(')')) {
    // The W3C spec also allows `feature=self` shorthand without
    // parens. Treat as self-only.
    if (value.toLowerCase() === 'self') {
      return { feature, allowlistRaw: 'self', isAllowAll: false, isDeny: false, hasSelf: true, origins: [] };
    }
    return null;
  }
  const inner = value.slice(1, -1).trim();
  if (inner === '') {
    return { feature, allowlistRaw: '', isAllowAll: false, isDeny: true, hasSelf: false, origins: [] };
  }
  // Tokenise on whitespace; quoted strings are origins.
  const tokens = inner.split(/\s+/);
  let hasSelf = false;
  const origins: string[] = [];
  let allowAll = false;
  for (const t of tokens) {
    if (t === '*') {
      allowAll = true;
    } else if (t.toLowerCase() === 'self') {
      hasSelf = true;
    } else if ((t.startsWith('"') && t.endsWith('"')) || (t.startsWith("'") && t.endsWith("'"))) {
      origins.push(t.slice(1, -1));
    } else {
      // Bare origin (non-spec but common in the wild).
      origins.push(t);
    }
  }
  return { feature, allowlistRaw: inner, isAllowAll: allowAll, isDeny: false, hasSelf, origins };
}

function getHeader(headers: Record<string, string> | undefined, name: string): string | null {
  if (!headers) return null;
  for (const [k, v] of Object.entries(headers)) {
    if (k.toLowerCase() === name) return v;
  }
  return null;
}

export function buildPermissionsPolicySnapshot(
  pageUrl: string,
  headers: Record<string, string> | undefined,
): PermissionsPolicySnapshot {
  const pageIsLocalhost = isLocalhost(pageUrl);
  const raw = getHeader(headers, 'permissions-policy');
  if (raw === null) {
    return { pageUrl, pageIsLocalhost, raw: null, directives: [], unparseable: false };
  }
  // Comma-split is correct for Permissions-Policy: per W3C the
  // allowlist itself never contains a comma at top level
  // (origins are quoted+space-separated). Empty directives are
  // tolerated by browsers (e.g. trailing comma).
  const parts = raw.split(',').map((s) => s.trim()).filter(Boolean);
  const directives: ParsedDirective[] = [];
  for (const p of parts) {
    const d = parseDirective(p);
    if (d) directives.push(d);
  }
  // Header present but no directive parsed → garbage.
  const unparseable = parts.length > 0 && directives.length === 0;
  return { pageUrl, pageIsLocalhost, raw, directives, unparseable };
}

export function detectPermissionsPolicyIssues(
  snap: PermissionsPolicySnapshot,
): PermissionsPolicyFinding[] {
  if (snap.pageIsLocalhost) return [];
  const out: PermissionsPolicyFinding[] = [];

  if (snap.raw === null) {
    out.push({
      severity: 'warn',
      kind: 'permissions-policy.missing',
      detail: `No Permissions-Policy header on this page. Every browser API (camera, microphone, geolocation, payment, USB, serial, MIDI, accelerometer, etc.) defaults to '*' — every embedded iframe inherits ambient permission. Set the header to deny each high-risk feature you don't use, e.g. 'Permissions-Policy: camera=(), microphone=(), geolocation=(), payment=()'.`,
      evidence: {},
    });
    return out;
  }

  if (snap.unparseable) {
    out.push({
      severity: 'warn',
      kind: 'permissions-policy.invalid',
      detail: `Permissions-Policy header is set but couldn't be parsed into any valid directive. Browsers ignore unparseable values, so the header has no effect. Header value: '${snap.raw.slice(0, 200)}'.`,
      evidence: { raw: snap.raw.slice(0, 200) },
    });
    return out;
  }

  // Per-directive checks.
  for (const d of snap.directives) {
    if (!HIGH_RISK_FEATURES.has(d.feature)) continue;
    if (d.isAllowAll) {
      out.push({
        severity: 'strict',
        kind: `permissions-policy.allow-all-${d.feature}`,
        detail: `Permissions-Policy directive '${d.feature}=${d.allowlistRaw}' explicitly allows every embedded iframe to use a high-risk feature. Restrict to '${d.feature}=()' (denied) or '${d.feature}=(self)' (page-only) unless an embed genuinely needs it.`,
        evidence: { feature: d.feature, allowlist: d.allowlistRaw },
      });
    }
  }

  // Detect HIGH_RISK features OMITTED from the policy. Because
  // the browser default is `*`, an explicit policy that lists
  // some features but not (e.g.) `camera` leaves camera at `*`.
  // We surface this so the operator knows to be exhaustive.
  const declaredFeatures = new Set(snap.directives.map((d) => d.feature));
  const omittedHighRisk: string[] = [];
  for (const f of HIGH_RISK_FEATURES) {
    if (!declaredFeatures.has(f)) omittedHighRisk.push(f);
  }
  if (omittedHighRisk.length > 0 && snap.directives.length > 0) {
    out.push({
      severity: 'warn',
      kind: 'permissions-policy.high-risk-omitted',
      detail: `Permissions-Policy declares ${snap.directives.length} directive(s) but omits ${omittedHighRisk.length} high-risk feature(s) which therefore default to '*'. Add restrictions for: ${omittedHighRisk.join(', ')}.`,
      evidence: { omitted: omittedHighRisk, declaredCount: snap.directives.length },
    });
  }

  return out;
}

File: crates/crawler-detectors/src/permissions_policy.rs
//! `permissions_policy` — Permissions-Policy response-header audit.
//! T76 port of `src/permissionsPolicy.ts`.
//!
//! Permissions-Policy (W3C, formerly Feature-Policy) gates which
//! browser APIs the page AND any embedded iframes can use. Without
//! the header — or with overly permissive directives — third-party
//! scripts and iframes inherit ambient permission to access camera,
//! microphone, geolocation, payment, USB, serial, MIDI, motion
//! sensors (keystroke-recovery side channel), display capture, etc.
//!
//! Findings (kinds + severities mirror TS byte-for-byte):
//!
//!   * `permissions-policy.missing`             — warn
//!   * `permissions-policy.allow-all-<feature>` — strict (per high-risk feature)
//!   * `permissions-policy.invalid`             — warn
//!   * `permissions-policy.high-risk-omitted`   — warn
//!
//! AVP-2 invariants: `unsafe_code = "deny"`, pure detector, no I/O.

use crate::url_helpers::is_localhost;
use crate::{AxisFinding, AxisSeverity};
use serde::{Deserialize, Serialize};
use std::collections::{BTreeMap, BTreeSet};

/// One parsed `feature=allowlist` directive.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct ParsedDirective {
    /// Lowercased feature name.
    pub feature: String,
    /// Raw allowlist text inside the parens (or `*` for unparenned star).
    /// Empty string for `()` (deny).
    pub allowlist_raw: String,
    /// True iff the allowlist is `*` — every iframe allowed.
    pub is_allow_all: bool,
    /// True iff the allowlist is `()` — denied for everyone.
    pub is_deny: bool,
    /// True iff `self` appears in the allowlist.
    pub has_self: bool,
    /// Origins beyond `self` listed in the allowlist (unquoted).
    pub origins: Vec<String>,
}

/// Captured page state the detector consumes.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct PermissionsPolicySnapshot {
    /// Top-level page URL.
    pub page_url: String,
    /// Localhost / loopback exemption.
    pub page_is_localhost: bool,
    /// Raw header value, or `None`.
    pub raw: Option<String>,
    /// Parsed directives in declaration order.
    pub directives: Vec<ParsedDirective>,
    /// True iff the header was present but unparseable into ANY directive.
    pub unparseable: bool,
}

/// High-risk features that trigger the strict `allow-all-<feature>`
/// finding when set to `*`. Drawn from the W3C Permissions-Policy
/// spec's privacy-impacting features list + payment + device-access
/// APIs known to enable cross-origin fingerprinting + motion sensors
/// (TouchLogger / AccessLogger keystroke side channel on mobile).
fn high_risk_features() -> BTreeSet<&'static str> {
    [
        "camera",
        "microphone",
        "geolocation",
        "payment",
        "usb",
        "serial",
        "midi",
        "hid",
        "bluetooth",
        "accelerometer",
        "gyroscope",
        "magnetometer",
        "display-capture",
        "screen-wake-lock",
    ]
    .into_iter()
    .collect()
}

fn header_lookup<'a>(headers: &'a BTreeMap<String, String>, name: &str) -> Option<&'a String> {
    headers
        .iter()
        .find(|(k, _)| k.eq_ignore_ascii_case(name))
        .map(|(_, v)| v)
}

/// Parse one `feature=allowlist` directive.
///
/// Returns `None` on malformed input (no `=`, empty value).
fn parse_directive(raw: &str) -> Option<ParsedDirective> {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return None;
    }
    let eq = trimmed.find('=')?;
    if eq == 0 {
        return None;
    }
    let feature = trimmed[..eq].trim().to_ascii_lowercase();
    if feature.is_empty() {
        return None;
    }
    let value = trimmed[eq + 1..].trim();
    if value.is_empty() {
        return None;
    }

    if value == "*" {
        return Some(ParsedDirective {
            feature,
            allowlist_raw: "*".into(),
            is_allow_all: true,
            is_deny: false,
            has_self: false,
            origins: Vec::new(),
        });
    }
    if !(value.starts_with('(') && value.ends_with(')')) {
        if value.eq_ignore_ascii_case("self") {
            return Some(ParsedDirective {
                feature,
                allowlist_raw: "self".into(),
                is_allow_all: false,
                is_deny: false,
                has_self: true,
                origins: Vec::new(),
            });
        }
        return None;
    }
    let inner = value[1..value.len() - 1].trim();
    if inner.is_empty() {
        return Some(ParsedDirective {
            feature,
            allowlist_raw: String::new(),
            is_allow_all: false,
            is_deny: true,
            has_self: false,
            origins: Vec::new(),
        });
    }
    let mut has_self = false;
    let mut origins = Vec::new();
    let mut allow_all = false;
    for t in inner.split_whitespace() {
        if t == "*" {
            allow_all = true;
        } else if t.eq_ignore_ascii_case("self") {
            has_self = true;
        } else if (t.starts_with('"') && t.ends_with('"'))
            || (t.starts_with('\'') && t.ends_with('\''))
        {
            origins.push(t[1..t.len() - 1].to_owned());
        } else {
            // Bare origin (non-spec but common in the wild).
            origins.push(t.to_owned());
        }
    }
    Some(ParsedDirective {
        feature,
        allowlist_raw: inner.to_owned(),
        is_allow_all: allow_all,
        is_deny: false,
        has_self,
        origins,
    })
}

/// Build a snapshot from a captured headers map.
pub fn build_permissions_policy_snapshot(
    page_url: &str,
    headers: &BTreeMap<String, String>,
) -> PermissionsPolicySnapshot {
    let page_is_localhost = is_localhost(page_url);
    let raw = header_lookup(headers, "permissions-policy").cloned();
    let Some(raw_str) = raw.clone() else {
        return PermissionsPolicySnapshot {
            page_url: page_url.to_owned(),
            page_is_localhost,
            raw: None,
            directives: Vec::new(),
            unparseable: false,
        };
    };

    let parts: Vec<&str> = raw_str
        .split(',')
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .collect();
    let directives: Vec<ParsedDirective> =
        parts.iter().filter_map(|p| parse_directive(p)).collect();
    let unparseable = !parts.is_empty() && directives.is_empty();

    PermissionsPolicySnapshot {
        page_url: page_url.to_owned(),
        page_is_localhost,
        raw: Some(raw_str),
        directives,
        unparseable,
    }
}

fn slice_for_log(s: &str, n: usize) -> String {
    s.chars().take(n).collect()
}

/// Pure detector: snapshot → findings. No I/O.
pub fn detect_permissions_policy_issues(snap: &PermissionsPolicySnapshot) -> Vec<AxisFinding> {
    if snap.page_is_localhost {
        return Vec::new();
    }
    let mut out = Vec::new();

    if snap.raw.is_none() {
        out.push(AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "permissions-policy.missing".into(),
            detail: "No Permissions-Policy header on this page. Every browser API (camera, microphone, geolocation, payment, USB, serial, MIDI, accelerometer, etc.) defaults to '*' — every embedded iframe inherits ambient permission. Set the header to deny each high-risk feature you don't use, e.g. 'Permissions-Policy: camera=(), microphone=(), geolocation=(), payment=()'.".into(),
        });
        return out;
    }

    if snap.unparseable {
        let raw_snippet = slice_for_log(snap.raw.as_deref().unwrap_or(""), 200);
        out.push(AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "permissions-policy.invalid".into(),
            detail: format!(
                "Permissions-Policy header is set but couldn't be parsed into any valid directive. Browsers ignore unparseable values, so the header has no effect. Header value: '{raw_snippet}'."
            ),
        });
        return out;
    }

    let high_risk = high_risk_features();

    for d in &snap.directives {
        if !high_risk.contains(d.feature.as_str()) {
            continue;
        }
        if d.is_allow_all {
            out.push(AxisFinding {
                severity: AxisSeverity::Strict,
                kind: format!("permissions-policy.allow-all-{}", d.feature),
                detail: format!(
                    "Permissions-Policy directive '{}={}' explicitly allows every embedded iframe to use a high-risk feature. Restrict to '{}=()' (denied) or '{}=(self)' (page-only) unless an embed genuinely needs it.",
                    d.feature, d.allowlist_raw, d.feature, d.feature
                ),
            });
        }
    }

    let declared: BTreeSet<&str> = snap.directives.iter().map(|d| d.feature.as_str()).collect();
    let omitted: Vec<&str> = high_risk
        .iter()
        .filter(|f| !declared.contains(*f))
        .copied()
        .collect();
    if !omitted.is_empty() && !snap.directives.is_empty() {
        out.push(AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "permissions-policy.high-risk-omitted".into(),
            detail: format!(
                "Permissions-Policy declares {} directive(s) but omits {} high-risk feature(s) which therefore default to '*'. Add restrictions for: {}.",
                snap.directives.len(),
                omitted.len(),
                omitted.join(", ")
            ),
        });
    }

    out
}

#[cfg(test)]
mod tests {
    use super::*;

    fn page() -> &'static str {
        "https://example.com/"
    }

    fn build(headers: &[(&str, &str)]) -> PermissionsPolicySnapshot {
        let map: BTreeMap<String, String> = headers
            .iter()
            .map(|(k, v)| ((*k).to_owned(), (*v).to_owned()))
            .collect();
        build_permissions_policy_snapshot(page(), &map)
    }

    #[test]
    fn localhost_skipped() {
        let map = BTreeMap::new();
        let s = build_permissions_policy_snapshot("http://localhost/", &map);
        assert!(detect_permissions_policy_issues(&s).is_empty());
    }

    #[test]
    fn missing_header_warns() {
        let s = build(&[]);
        let f = detect_permissions_policy_issues(&s);
        assert!(f.iter().any(|x| x.kind == "permissions-policy.missing"));
    }

    #[test]
    fn parses_deny_directive() {
        let s = build(&[("Permissions-Policy", "camera=()")]);
        assert_eq!(s.directives.len(), 1);
        assert!(s.directives[0].is_deny);
    }

    #[test]
    fn parses_self_only_directive() {
        let s = build(&[("Permissions-Policy", "camera=(self)")]);
        assert!(s.directives[0].has_self);
        assert!(!s.directives[0].is_allow_all);
    }

    #[test]
    fn parses_star_directive() {
        let s = build(&[("Permissions-Policy", "camera=*")]);
        assert!(s.directives[0].is_allow_all);
    }

    #[test]
    fn allow_all_high_risk_is_strict() {
        let s = build(&[("Permissions-Policy", "camera=*")]);
        let f = detect_permissions_policy_issues(&s);
        assert!(f.iter().any(|x| {
            x.kind == "permissions-policy.allow-all-camera" && x.severity == AxisSeverity::Strict
        }));
    }

    #[test]
    fn allow_all_low_risk_does_not_fire_strict() {
        // 'fullscreen' is not in the HIGH_RISK set.
        let s = build(&[(
            "Permissions-Policy",
            "fullscreen=*, camera=(), microphone=(), geolocation=(), payment=(), usb=(), serial=(), midi=(), hid=(), bluetooth=(), accelerometer=(), gyroscope=(), magnetometer=(), display-capture=(), screen-wake-lock=()",
        )]);
        let f = detect_permissions_policy_issues(&s);
        assert!(!f
            .iter()
            .any(|x| x.kind.starts_with("permissions-policy.allow-all-")));
    }

    #[test]
    fn unparseable_warns() {
        // Header is non-empty after split, but no = signs → no parseable directives
        let s = build(&[("Permissions-Policy", "garbage, also-garbage")]);
        let f = detect_permissions_policy_issues(&s);
        assert!(f.iter().any(|x| x.kind == "permissions-policy.invalid"));
    }

    #[test]
    fn high_risk_omitted_warns() {
        let s = build(&[("Permissions-Policy", "camera=()")]);
        let f = detect_permissions_policy_issues(&s);
        assert!(f
            .iter()
            .any(|x| x.kind == "permissions-policy.high-risk-omitted"));
    }

    #[test]
    fn fully_locked_down_policy_is_clean() {
        let s = build(&[(
            "Permissions-Policy",
            "camera=(), microphone=(), geolocation=(), payment=(), usb=(), serial=(), midi=(), hid=(), bluetooth=(), accelerometer=(), gyroscope=(), magnetometer=(), display-capture=(), screen-wake-lock=()",
        )]);
        let f = detect_permissions_policy_issues(&s);
        assert!(
            f.is_empty(),
            "fully-locked policy should be silent, got: {f:#?}"
        );
    }

    #[test]
    fn parses_quoted_origin_in_allowlist() {
        let s = build(&[("Permissions-Policy", "camera=(self \"https://e.com\")")]);
        assert!(s.directives[0].has_self);
        assert_eq!(s.directives[0].origins, vec!["https://e.com".to_string()]);
    }

    #[test]
    fn parses_self_shorthand_without_parens() {
        let s = build(&[("Permissions-Policy", "camera=self")]);
        assert!(s.directives[0].has_self);
    }

    #[test]
    fn header_lookup_case_insensitive() {
        let s = build(&[("PERMISSIONS-POLICY", "camera=()")]);
        assert_eq!(s.directives.len(), 1);
    }

    #[test]
    fn snapshot_round_trips_through_serde() {
        let s = build(&[("Permissions-Policy", "camera=(), microphone=(self)")]);
        let j = serde_json::to_string(&s).expect("ser");
        let back: PermissionsPolicySnapshot = serde_json::from_str(&j).expect("de");
        assert_eq!(back.directives.len(), s.directives.len());
    }

    #[test]
    fn allow_all_microphone_strict_too() {
        // SECURITY: every high-risk feature must fire individually.
        for feat in [
            "microphone",
            "geolocation",
            "payment",
            "usb",
            "serial",
            "midi",
            "hid",
            "bluetooth",
            "accelerometer",
            "gyroscope",
            "magnetometer",
            "display-capture",
            "screen-wake-lock",
        ] {
            let s = build(&[("Permissions-Policy", &format!("{feat}=*"))]);
            let f = detect_permissions_policy_issues(&s);
            assert!(
                f.iter().any(|x| {
                    x.kind == format!("permissions-policy.allow-all-{feat}")
                        && x.severity == AxisSeverity::Strict
                }),
                "feature {feat} must fire strict allow-all finding"
            );
        }
    }
}
